tests/gradients/test_gradient_transform.py

```diff
@@ -16,7 +16,78 @@
+    reduced copy of the eponymous method of ``JacobianTape``."""
+
+    @pytest.mark.parametrize("method", ["analytic", "best"])
+    def test_with_nondiff_parameters(self, method):
+        """Test that trainable parameters without grad_method
+        are detected correctly, raising an exception."""
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(np.array(0.1, requires_grad=True), wires=0)
+            qml.RX(np.array(0.1, requires_grad=True), wires=0)
+            qml.expval(qml.PauliZ(0))
+        tape._par_info[0]["grad_method"] = "A"
+        tape._par_info[1]["grad_method"] = None
+        with pytest.raises(ValueError, match="Cannot differentiate with respect"):
+            _grad_method_validation(method, tape)
+
+    def test_with_numdiff_parameters_and_analytic(self):
+        """Test that trainable parameters with numerical grad_method ``"F"``
+        together with ``method="analytic"`` raises an exception."""
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(np.array(0.1, requires_grad=True), wires=0)
+            qml.RX(np.array(0.1, requires_grad=True), wires=0)
+            qml.expval(qml.PauliZ(0))
+        tape._par_info[0]["grad_method"] = "A"
+        tape._par_info[1]["grad_method"] = "F"
+        with pytest.raises(ValueError, match="The analytic gradient method cannot be used"):
+            _grad_method_validation("analytic", tape)
+
+
+class TestChooseParamsWithMethods:
+    """Test the helper function _choose_params_with_methods, which is a
+    reduced copy of the eponymous method of ``JacobianTape``."""
+
+    all_diff_methods = [
+        ["A"] * 2,
+        [None, "A", "F", "A"],
+        ["F"],
+        ["0", "A"],
+    ]
+
+    @pytest.mark.parametrize("diff_methods", all_diff_methods)
+    def test_without_argnum(self, diff_methods):
+        """Test that the method returns all diff_methods when
+        used with ``argnum=None``."""
+        chosen = _choose_params_with_methods(diff_methods, None)
+        assert chosen == dict(enumerate(diff_methods))
+
+    @pytest.mark.parametrize(
+        "diff_methods, argnum, expected",
+        zip(all_diff_methods, [1, 2, 0, 0], [{1: "A"}, {2: "F"}, {0: "F"}, {0: "0"}]),
+    )
+    def test_with_integer_argnum(self, diff_methods, argnum, expected):
+        """Test that the method returns the correct single diff_method when
+        used with an integer ``argnum``."""
+        chosen = _choose_params_with_methods(diff_methods, argnum)
+        assert chosen == expected
+
+    @pytest.mark.parametrize("diff_methods", all_diff_methods[:2] + [[]])
+    def test_warning_with_empty_argnum(self, diff_methods):
+        """Test that the method raises a warning when an empty iterable
+        is passed as ``argnum``."""
+        with pytest.warns(UserWarning, match="No trainable parameters were specified"):
+            chosen = _choose_params_with_methods(diff_methods, [])
+        assert chosen == {}

```

Comment:
nice tests 💪 

I'm curious, were there existing tests for these methods in the `JacobianTape` test file? Are these those tests?

Reply:
For `_choose_params_with_methods`, this basically is a duplicate of corresponding tests in `tests/tape/test_jacobian_tape.py` , yes.
`JacobianTape._grad_method_validation` does not seem to have dedicated unit tests, these are newly written.

Reply:
💯